File: 2 Add Two Numbers/solution.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun addTwoNumbers(l1: ListNode?, l2: ListNode?): ListNode? {
        var n1 = l1;
        var n2 = l2;
        var dummyHead = ListNode(0);
        var current = dummyHead;
        var carry = 0;

        while (n1 != null || n2 != null || carry != 0) {
            val v1 = if (n1 != null) n1.`val` else 0;
            val v2 = if (n2 != null) n2.`val` else 0;
            val sum = carry + v1 + v2;
            carry = sum / 10;
            current.next = ListNode(sum % 10);
            current = current.next;
            n1 = if (n1 != null) n1.next else null;
            n2 = if (n2 != null) n2.next else null;
        }

        return dummyHead.next;
    }
}

File: 2 Add Two Numbers/solution.dart
/**
 * Definition for singly-linked list.
 * class ListNode {
 *   int val;
 *   ListNode? next;
 *   ListNode([this.val = 0, this.next]);
 * }
 */
class Solution {
  ListNode? addTwoNumbers(ListNode? l1, ListNode? l2) {
    var dummyHead = new ListNode(0);
    var current = dummyHead;
    var carry = 0;

    while (l1 != null || l2 != null || carry != 0) {
        var v1 = l1 != null ? l1.val : 0;
        var v2 = l2 != null ? l2.val : 0;
        var sum = carry + v1 + v2;
        carry = (sum / 10).toInt();
        current.next = new ListNode(sum % 10);
        current = current.next!;
        l1 = l1 != null ? l1.next : null;
        l2 = l2 != null ? l2.next : null;
    }

    return dummyHead.next;
  }
}